src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -0,0 +1,51 @@
+require 'puppet_pal'
+require 'puppet/server/log_collector'
+require 'puppet/server/logging'
+
+module Puppet
+  module Server
+    class ASTCompiler
+      def self.compile(compile_options)
+        options = compile_options['options'] || {}
+
+        log_level = options['log_level']
+
+        if options['capture_logs']
+          catalog, logs = Logging.capture_logs(log_level) do
+            compile_ast(compile_options)
+          end
+
+          { catalog: catalog, logs: logs }
+        else
+          catalog = compile_ast(compile_options)
+          { catalog: catalog }
+        end
+      end
+
+      def self.compile_ast(compile_options)
+        # FromDataConverter.convert below expects this to be a hash
+        code = JSON.parse(compile_options['code_ast'])
+        target = compile_options['certname']
+
+        # Use the existing environment with the requested name
+        Puppet::Pal.in_environment(compile_options['environment'],
+                                   envpath: Puppet[:environmentpath],
+                                   facts: compile_options['facts']['values'],
+                                   variables: compile_options['variables']['values']) do |pal|
+          Puppet.lookup(:pal_current_node).trusted_data = compile_options['trusted_facts']['values']
+
+          # This compiler has been configured with a node containing
+          # the requested environment, facts, and variables, and is used
+          # to compile a catalog in that context from the supplied AST.
+          pal.with_catalog_compiler do |compiler|
+            ast = Puppet::Pops::Serialization::FromDataConverter.convert(code)
```

Comment:
Does the deserialization need to happen within this block?

Reply:
So apparently something gets loaded as part of setting up these blocks that allows the `FromDataConverter` to return a `Puppet::Pops::Model::Program`, which lets us follow the right branch [here](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/pal/catalog_compiler.rb#L48). Doing it outside the block returns what looks like an anonymous class. I'll try to track down what needs to be loaded so we could do it outside, if that seems more appropriate.

Reply:
I don't know that it needs to be deserialized outside of the block, though I find that behavior surprising. It would be nice to somehow make the dependency on where it's executed explicit if not remove it. What do you think?

Reply:
Yeah I do think that would be better, unless the required loading proves complicated. I get the impression some types just need to be brought into scope.

Reply:
Turns out that it requires the Loaders to be initialized with the environment and then added to the context, where they are used to load various PCore types. Without this, a generic class is generated instead, in place of the official PCore type. I think it's best to leave that initialization up to PAL, and therefore do the AST conversion inside this block.